File: Laboratory work #3/Python/ZeidelAndJakoby.py
import numpy as np
import math

class ZeidelAndJakoby :
    def __init__ (self) :
        self.systemSize = 0
        self.coeff = []
        self.freeTerm = []
        self.filename = "mat.txt"
        self.eps = 0.001

    def initializationFromFile(self) :
        with open(self.filename) as file :
            self.coeff = [list(map(float, row.split())) for row in file.readlines()]
            self.systemSize = len(self.coeff[0])-1

        for i in range(0,self.systemSize) :
            self.freeTerm.append([self.coeff[i][-1]])
            self.coeff[i].pop()

    def showSystem(self) :
        print("Р’РІРµРґРµРЅРЅР°СЏ СЃРёСЃС‚РµРјР°: ")
        for row in range(len(self.freeTerm)):
            for col in range(len(self.coeff[row])):
                print(f"{self.coeff[row][col]}*x{col}", end='')
                if col < self.systemSize - 1 :
                    print(" + ", end='')
            print(f" = {self.freeTerm[row][0]}")


    def applyJakobyMethod(self) :
        print("--------JAKOBY--------")
        npCoeff = np.asarray(self.coeff)
        npFreeTerm = np.asarray(self.freeTerm)

        tmp = []
        for i in range(self.systemSize) :
            tmp.append([])
            for j in range(self.systemSize) :
                tmp[i].append(0)
            tmp[i][i] = self.coeff[i][i]

        diagonalMatrix = np.asarray(tmp)
        inversedDiagMatrix = np.linalg.inv(diagonalMatrix)

        # print(npCoeff)
        # print(npFreeTerm)
        # print(inversedDiagMatrix)

        b = np.eye(self.systemSize) - np.dot(inversedDiagMatrix,npCoeff)
        d = np.dot(inversedDiagMatrix,npFreeTerm)

        currentX = np.array([[0]]*self.systemSize,dtype = float)
        previousX = np.array([[1]]*self.systemSize,dtype = float)

        iterations = 0
        while True :
            currentX = np.dot(b,previousX) + d

            diff = currentX - previousX

            maxValue = math.fabs(diff.max())
            minValue = math.fabs(diff.min())

            iterations += 1
            if maxValue > minValue and maxValue < self.eps :
                print(f"Number of iterations: {iterations}")
                for i in range(self.systemSize) :
                    print(f"x[{i}] = {currentX[i]}")
                break;
            elif maxValue < minValue and minValue < self.eps :
                print(f"Number of iterations: {iterations}")
                for i in range(self.systemSize) :
                    print(f"x[{i}] = {currentX[i]}")
                break;

            previousX = currentX

    def applyZeidelMethod(self) :
        print("--------ZEIDEL--------")
        npCoeff = np.asarray(self.coeff)
        npFreeTerm = np.asarray(self.freeTerm)

        triangMatrix = np.tril(npCoeff)
        inversedMatrix = np.linalg.inv(triangMatrix)

        b = np.eye(self.systemSize) - np.dot(inversedMatrix,npCoeff)
        d = np.dot(inversedMatrix,npFreeTerm)

        currentX = np.array([[0]]*self.systemSize,dtype = float)
        previousX = np.array([[1]]*self.systemSize,dtype = float)

        iterations = 0
        while True :
            currentX = np.dot(b,previousX) + d

            diff = currentX - previousX

            maxValue = math.fabs(diff.max())
            minValue = math.fabs(diff.min())

            iterations += 1
            if maxValue > minValue and maxValue < self.eps :
                print(f"Number of iterations: {iterations}")
                for i in range(self.systemSize) :
                    print(f"x[{i}] = {currentX[i]}")
                break;
            elif maxValue < minValue and minValue < self.eps :
                print(f"Number of iterations: {iterations}")
                for i in range(self.systemSize) :
                    print(f"x[{i}] = {currentX[i]}")
                break;

            previousX = currentX

if __name__ == '__main__':
    solve = ZeidelAndJakoby()

    solve.initializationFromFile()
    solve.showSystem()
    solve.applyJakobyMethod()
    solve.applyZeidelMethod(

File: Laboratory work #3/Matlab/ZeidelAndJakoby.m
classdef ZeidelAndJakoby < handle
    properties
        systemSize = 0;
        coeff = [];
        freeTerm = [];
        filename = "mat.txt";
        eps = 0.001
    end
        
     methods        
        function initializationFromFile(obj)
            m = 0;
            n = 0;
            
            obj.coeff = load(obj.filename);
            [m,n] = size(obj.coeff);
            
            obj.systemSize = m;

            for i = 1:obj.systemSize
               obj.freeTerm(i,1) = obj.coeff(i,obj.systemSize+1);
               obj.coeff(i,obj.systemSize+1) = 0;
            end
            obj.coeff(:,obj.systemSize+1)=[];
        end
        
        function showSystem(obj)
            disp('Введённая система:');
            for i = 1:obj.systemSize
                for j = 1:obj.systemSize
                    if (obj.coeff(i,j) < 0)
                        fprintf('(%f)*x%d',obj.coeff(i,j), j);
                    else
                        fprintf('%f*x%d',obj.coeff(i,j), j);
                    end
                    
                    if (j < obj.systemSize)
                        fprintf( ' + ' );
                    end 
                end
                fprintf(' = %d\n', obj.freeTerm(i));
            end
        end
        
        function applyJakobyMethod(obj)
            disp("--------JAKOBY--------");
            diagonalElements = diag(obj.coeff);
            diagonalMatrix = diag(diagonalElements);
            inversedDiagMatrix = inv(diagonalMatrix);
            
            b = eye(obj.systemSize) - mtimes(inversedDiagMatrix,obj.coeff);
            d = mtimes(inversedDiagMatrix,obj.freeTerm);
            
            currentX = zeros(obj.systemSize,1);
            previousX = ones(obj.systemSize,1);
            
            iterations = 0;
            while 1
               currentX = mtimes(b,previousX) + d;
               diff = currentX - previousX;
               
               maxValue = abs(max(diff));
               minValue = abs(min(diff));
               
               iterations = iterations + 1; 
               if ( (maxValue > minValue) && (maxValue < obj.eps) )
                   fprintf ('Iterations:   %d\n',iterations);
                   for i = 1:obj.systemSize
                       fprintf ('x[%d] = %f\n',i, currentX(i));
                   end
                   break;
               elseif ( (maxValue < minValue) && (minValue < obj.eps) )
                   fprintf ('Iterations:   %d\n',iterations);
                   for i = 1:obj.systemSize
                       fprintf ('x[%d] = %f\n',i, currentX(i));
                   end
                   break;
               end
               
               previousX = currentX;
            end
            
        end

        function applyZeidelMethod(obj)
            disp("--------ZEIDEL--------");
            triangMatrix = tril(obj.coeff);
            inversedMatrix = inv(triangMatrix);
            
            b = eye(obj.systemSize) - mtimes(inversedMatrix,obj.coeff);
            d = mtimes(inversedMatrix,obj.freeTerm);
            
            currentX = zeros(obj.systemSize,1);
            previousX = ones(obj.systemSize,1);
            
            iterations = 0;
            while 1
               currentX = mtimes(b,previousX) + d;
               diff = currentX - previousX;
               
               maxValue = abs(max(diff));
               minValue = abs(min(diff));
               
               iterations = iterations + 1; 
               if ( (maxValue > minValue) && (maxValue < obj.eps) )
                   fprintf ('Iterations:   %d\n',iterations);
                   for i = 1:obj.systemSize
                       fprintf ('x[%d] = %f\n',i, currentX(i));
                   end
                   break;
               elseif ( (maxValue < minValue) && (minValue < obj.eps) )
                   fprintf ('Iterations:   %d\n',iterations);
                   for i = 1:obj.systemSize
                       fprintf ('x[%d] = %f\n',i, currentX(i));
                   end
                   break;
               end
               
               previousX = currentX;
            end
        end
        
    end
end